README.rst
flask-oidc
==========

`OpenID Connect <https://openid.net/connect/>`_ support for `Flask <http://flask.pocoo.org/>`_.

.. image:: https://img.shields.io/pypi/v/flask-oidc.svg?style=flat
   :target: https://pypi.python.org/pypi/flask-oidc
   :alt: PyPI version

.. image:: https://img.shields.io/pypi/dm/flask-oidc.svg?style=flat
   :target: https://pypi.python.org/pypi/flask-oidc
   :alt: Downloads per month

.. image:: https://readthedocs.org/projects/flask-oidc/badge/?version=latest
   :target: http://flask-oidc.readthedocs.io/en/latest/?badge=latest
   :alt: Documentation Status

.. image:: https://github.com/fedora-infra/flask-oidc/actions/workflows/main.yml/badge.svg?branch=develop
   :target: https://github.com/fedora-infra/flask-oidc/actions/workflows/main.yml?query=branch%3Adevelop
   :alt: Tests Status


This library should work with any standards compliant OpenID Connect provider.

It has been tested with:

* `Ipsilon <https://ipsilon-project.org/>`_


Project status
==============

This project is in active development (again).


docs/changelog.md
# Changelog


## 2.1.1 (2023-10-27)

### Fixed

- Avoid redirect loops when the app is not mounted on the webserver root (#52)

### Changed

- Use REUSE for licences
- Convert the changelog to markdown

### Added

- Publish to PyPI and Github when a tag is pushed


## 2.1.0 (2023-10-09)

### Fixed

- Handle token expiration when there is no `refresh_token` or no token URL (#39)

### Changed

- Restore the `OVERWRITE_REDIRECT_URI` configuration option as
  `OIDC_OVERWRITE_REDIRECT_URI`.
- The `redirect_uri` that is generated and sent to the ID provider is no longer
  forced to HTTPS, because the
  [the OIDC spec](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest)
  is actually only a strong recommendation (#35). You can
  use `OVERWRITE_REDIRECT_URI` if you want to force it to HTTPS (or any other
  URL).


## 2.0.3 (2023-09-08)

### Fixed

- Use the `OIDC_CALLBACK_ROUTE` with the ID provider when it is defined,
  instead of the default (#21)
- Auto-renew tokens when they have expired (if possible), as version 1.x used
  to do (#19)

### Changed

- The `redirect_uri` that is generated and sent to the ID provider is always
  HTTPS, as [the OIDC spec](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest)
  mandates.
- Don't request the `profile` scope by default, as version 1.x used to do
  (#21).


## 2.0.2 (2023-08-23)

### Fixed

- Avoid a redirect loop on logout when the token is expired (#17).


### Deprecated

- Configuration option `OIDC_USERINFO_URL` (and the `userinfo_uri` key in
  `client_secrets`) (#15).


## 2.0.1 (2023-08-22)

This is a bugfix release.

### Fixed

- Don't crash if the `client_secrets` don't contain a `userinfo_uri` key (#13).
- Handle older versions of Werkzeug.


## 2.0.0 (2023-08-21)

This is a major release that rebases the Flask OIDC API on
[Authlib](https://authlib.org/).

### Removed

- Custom callback with the `OpenIDConnect.custom_callback()` decorator
- Registration has been moved to the
  [oidc-register package](https://pypi.org/project/oidc-register/)
- Configuration option `OIDC_GOOGLE_APPS_DOMAIN`
- Configuration option `OIDC_VALID_ISSUERS`
- Configuration option `OIDC_REQUIRE_VERIFIED_EMAIL`
- Configuration option `OIDC_RESOURCE_CHECK_AUD`
- The following parameters of the `OpenIDConnect` constructor have been
  removed:

  - `credentials_store`
  - `http`
  - `time`
  - `urandom`

### Deprecated

- Configuration option `OIDC_OPENID_REALM`
- Configuration option `OIDC_CALLBACK_ROUTE`
- Configuration option `OVERWRITE_REDIRECT_URI`
- The following configuration options have been removed because the
  functionality is now handled by Authlib:

  - `OIDC_ID_TOKEN_COOKIE_NAME`
  - `OIDC_ID_TOKEN_COOKIE_PATH`
  - `OIDC_ID_TOKEN_COOKIE_TTL`
  - `OIDC_COOKIE_SECURE`

- The `OpenIDConnect.user_getinfo()` and `OpenIDConnect.user_getfield()`
  methods are deprecated, you'll find all the user information in the
  session: `session["oidc_auth_profile"]`.
  If you need to get the user information using a specific token, you can
  do so by calling `g._oidc_auth.userinfo(token=token)`.
- The `OpenIDConnect.logout()` method is deprecated, just redirect to the
  `/logout` view.

### Changed

The callback route (aka "redirect URL") is not configurable with
`OIDC_CALLBACK_ROUTE` anymore. It is always `/authorize`, but a prefix can
be configured when instanciating the `OpenIDConnect` extension (or calling
its `OpenIDConnect.init_app()` method:

```python
    app = Flask(__name__)
    openid = OpenIDConnect(app, prefix="/oidc")
    # The OIDC redirect URL will be /oidc/authorize
```

This will also give you `/login` and `/logout` views, prefixed identically.

The `OIDC_SCOPES` configuration value should now be a string, where the
scopes are separated with spaces.

The minimum Python version is `3.8`.

### Added

The `OpenIDConnect.accept_token()` decorator now accepts a `scopes` parameter,
which is a list of scopes that the provided token must include for the view to
be authorized. It is an Authlib
[ResourceProtector](https://docs.authlib.org/en/latest/flask/2/resource-server.html).

The Authlib app is available in the `g._oidc_auth` variable. This means that
there cannot be more than one `OpenIDConnect` extension on a given Flask
application. If you need more, we advise you to use Authlib directly.

### Development

- A [pre-commit](https://pre-commit.com/) config has been added, please enable
  it with `pre-commit install`
- Unit tests are run by [Tox](https://tox.readthedocs.io/)
- A coverage report is produced, and the coverage must be 100%
- The git submodule for documentation themes has been dropped
- The code is formatted using [Black](https://black.readthedocs.io/)
- The code is linted using [Ruff](https://ruff.rs)
- The package metadata is managed by [Poetry](https://python-poetry.org/)
- CI is run using Github Actions, Travis config has been dropped
- Dependencies are updated using [Renovate](https://docs.renovatebot.com/)


tests/app.py
# SPDX-FileCopyrightText: 2014-2015 Erica Ehrhardt
# SPDX-FileCopyrightText: 2016-2022 Patrick Uiterwijk <patrick@puiterwijk.org>
# SPDX-FileCopyrightText: 2023 Aurélien Bompard <aurelien@bompard.org>
#
# SPDX-License-Identifier: BSD-2-Clause

"""
Flask app for testing the OpenID Connect extension.
"""

import json

from flask import Blueprint, Flask, g

from flask_oidc import OpenIDConnect

oidc = OpenIDConnect()
bp = Blueprint("main", __name__)


@bp.route("/")
@oidc.require_login
def index():
    return "too many secrets", 200, {"Content-Type": "text/plain; charset=utf-8"}


@bp.route("/at")
@oidc.require_login
def get_at():
    return oidc.get_access_token(), 200, {"Content-Type": "text/plain; charset=utf-8"}


@bp.route("/rt")
@oidc.require_login
def get_rt():
    return oidc.get_refresh_token(), 200, {"Content-Type": "text/plain; charset=utf-8"}


@oidc.require_login
def raw_api():
    return {"token": g.oidc_token_info}


@bp.route("/api", methods=["GET", "POST"])
@oidc.require_login
def api():
    return json.dumps(raw_api())


@bp.route("/need-token")
@oidc.accept_token()
def need_token():
    return "OK"


@bp.route("/need-profile")
@oidc.accept_token(scopes=["profile"])
def need_profile():
    return "OK"


def create_app(config, oidc_overrides=None):
    oidc_overrides = oidc_overrides or {}
    app = Flask(__name__)
    app.config.update(config)
    oidc.init_app(app, **oidc_overrides)
    # useful for tests
    app.oidc_ext = oidc

    app.register_blueprint(bp)
    # Check combination with an external API renderer like Flask-RESTful

    def externally_rendered_api(*args, **kwds):
        inner_response = raw_api(*args, **kwds)
        if isinstance(inner_response, tuple):
            raw_response, response_code, headers = inner_response
            rendered_response = json.dumps(raw_response), response_code, headers
        else:
            rendered_response = json.dumps(inner_response)
        return rendered_response

    app.add_url_rule(
        "/external_api", view_func=externally_rendered_api, methods=["GET", "POST"]
    )

    return app


tests/__init__.py
# SPDX-FileCopyrightText: 2014-2015 Erica Ehrhardt
# SPDX-FileCopyrightText: 2016-2022 Patrick Uiterwijk <patrick@puiterwijk.org>
# SPDX-FileCopyrightText: 2023 Aurélien Bompard <aurelien@bompard.org>
#
# SPDX-License-Identifier: BSD-2-Clause


tests/conftest.py
# SPDX-FileCopyrightText: 2014-2015 Erica Ehrhardt
# SPDX-FileCopyrightText: 2016-2022 Patrick Uiterwijk <patrick@puiterwijk.org>
# SPDX-FileCopyrightText: 2023 Aurélien Bompard <aurelien@bompard.org>
#
# SPDX-License-Identifier: BSD-2-Clause

import json
import time
from importlib.resources import path

import pytest
import responses

from . import app


@pytest.fixture
def dummy_token():
    return {
        "token_type": "Bearer",
        "access_token": "dummy_access_token",
        "refresh_token": "dummy_refresh_token",
        "expires_in": "3600",
        "expires_at": int(time.time()) + 3600,
    }


@pytest.fixture(scope="session")
def client_secrets_path():
    with path("tests", "client_secrets.json") as filepath:
        yield filepath.as_posix()


@pytest.fixture(scope="session")
def client_secrets(client_secrets_path):
    """The parsed contents of `client_secrets.json`."""
    with open(client_secrets_path) as f:
        return json.load(f)["web"]


@pytest.fixture
def mocked_responses():
    with responses.RequestsMock(assert_all_requests_are_fired=False) as rsps:
        yield rsps


@pytest.fixture(scope="session")
def oidc_server_metadata(client_secrets):
    """IdP server metadata used in tests."""
    base_url = client_secrets["issuer"].rstrip("/")
    return {
        "issuer": f"{base_url}/",
        "authorization_endpoint": f"{base_url}/Authorization",
        "token_endpoint": f"{base_url}/Token",
        "userinfo_endpoint": f"{base_url}/UserInfo",
        "introspection_endpoint": f"{base_url}/TokenInfo",
        # "jwks_uri": f"{base_url}/Jwks",
    }


@pytest.fixture
def make_test_app(
    oidc_server_metadata, client_secrets_path, client_secrets, mocked_responses
):
    """Make a Flask app object set up for testing."""

    def _make_test_app(config=None, oidc_overrides=None):
        _config = {
            "SECRET_KEY": "SEEEKRIT",
            "TESTING": True,
            "OIDC_CLIENT_SECRETS": client_secrets_path,
        }
        _config.update(config or {})

        test_app = app.create_app(_config, oidc_overrides or {})

        base_url = client_secrets["issuer"].rstrip("/")
        mocked_responses.get(
            f"{base_url}/.well-known/openid-configuration", json=oidc_server_metadata
        )
        return test_app

    return _make_test_app


@pytest.fixture
def test_app(make_test_app):
    """A Flask app object set up for testing."""
    return make_test_app()


@pytest.fixture
def test_client(test_app):
    """A Flask test client for the test app."""
    return test_app.test_client()


@pytest.fixture
def client(test_client):
    """A Flask test client for the test app."""
    with test_client:
        yield test_client


flask_oidc/views.py
# SPDX-FileCopyrightText: 2014-2015 Erica Ehrhardt
# SPDX-FileCopyrightText: 2016-2022 Patrick Uiterwijk <patrick@puiterwijk.org>
# SPDX-FileCopyrightText: 2023 Aurélien Bompard <aurelien@bompard.org>
#
# SPDX-License-Identifier: BSD-2-Clause

import logging
import warnings
from urllib.parse import urlparse

from authlib.integrations.base_client.errors import OAuthError
from flask import (
    Blueprint,
    abort,
    current_app,
    flash,
    g,
    redirect,
    request,
    session,
    url_for,
)

logger = logging.getLogger(__name__)

auth_routes = Blueprint("oidc_auth", __name__)


@auth_routes.route("/login", endpoint="login")
def login_view():
    if current_app.config["OIDC_OVERWRITE_REDIRECT_URI"]:
        redirect_uri = current_app.config["OIDC_OVERWRITE_REDIRECT_URI"]
    elif current_app.config["OIDC_CALLBACK_ROUTE"]:
        redirect_uri = (
            f"https://{request.host}{current_app.config['OIDC_CALLBACK_ROUTE']}"
        )
    else:
        redirect_uri = url_for("oidc_auth.authorize", _external=True)
    session["next"] = request.args.get("next", request.root_url)
    return g._oidc_auth.authorize_redirect(redirect_uri)


@auth_routes.route("/authorize", endpoint="authorize")
def authorize_view():
    try:
        token = g._oidc_auth.authorize_access_token()
    except OAuthError as e:
        logger.exception("Could not get the access token")
        abort(401, str(e))
    session["oidc_auth_token"] = token
    g.oidc_id_token = token
    if current_app.config["OIDC_USER_INFO_ENABLED"]:
        profile = g._oidc_auth.userinfo(token=token)
        session["oidc_auth_profile"] = profile
    try:
        return_to = session["next"]
        del session["next"]
    except KeyError:
        return_to = request.root_url
    return redirect(return_to)


@auth_routes.route("/logout", endpoint="logout")
def logout_view():
    """
    Request the browser to please forget the cookie we set, to clear the
    current session.

    Note that as described in [1], this will not log out in the case of a
    browser that doesn't clear cookies when requested to, and the user
    could be automatically logged in when they hit any authenticated
    endpoint.

    [1]: https://github.com/puiterwijk/flask-oidc/issues/5#issuecomment-86187023

    .. versionadded:: 1.0
    """
    session.pop("oidc_auth_token", None)
    session.pop("oidc_auth_profile", None)
    g.oidc_id_token = None
    reason = request.args.get("reason")
    if reason == "expired":
        flash("Your session expired, please reconnect.")
    else:
        flash("You were successfully logged out.")
    return_to = request.args.get("next", request.root_url)
    return redirect(return_to)


def legacy_oidc_callback():
    warnings.warn(
        "The {callback_url} route is deprecated, please use {authorize_url}".format(
            callback_url=current_app.config["OIDC_CALLBACK_ROUTE"],
            authorize_url=url_for("oidc_auth.authorize"),
        ),
        DeprecationWarning,
        stacklevel=2,
    )
    return redirect(
        "{url}?{qs}".format(
            url=url_for("oidc_auth.authorize"), qs=urlparse(request.url).query
        )
    )


flask_oidc/__init__.py
# SPDX-FileCopyrightText: 2014-2015 Erica Ehrhardt
# SPDX-FileCopyrightText: 2016-2022 Patrick Uiterwijk <patrick@puiterwijk.org>
# SPDX-FileCopyrightText: 2023 Aurélien Bompard <aurelien@bompard.org>
#
# SPDX-License-Identifier: BSD-2-Clause

import json
import logging
import warnings
from functools import wraps
from urllib.parse import quote_plus

from authlib.common.errors import AuthlibBaseError
from authlib.integrations.base_client import InvalidTokenError
from authlib.integrations.flask_client import OAuth
from authlib.integrations.flask_oauth2 import ResourceProtector
from authlib.oauth2.rfc6749 import OAuth2Token
from authlib.oauth2.rfc7662 import (
    IntrospectTokenValidator as BaseIntrospectTokenValidator,
)
from flask import abort, current_app, g, redirect, request, session, url_for

from .views import auth_routes, legacy_oidc_callback

__all__ = ["OpenIDConnect"]

_CONFIG_REMOVED = (
    "OIDC_GOOGLE_APPS_DOMAIN",
    "OIDC_REQUIRE_VERIFIED_EMAIL",
    "OIDC_RESOURCE_CHECK_AUD",
    "OIDC_VALID_ISSUERS",
)
_CONFIG_DEPRECATED = (
    "OIDC_ID_TOKEN_COOKIE_NAME",
    "OIDC_ID_TOKEN_COOKIE_PATH",
    "OIDC_ID_TOKEN_COOKIE_TTL",
    "OIDC_COOKIE_SECURE",
    "OIDC_OPENID_REALM",
    "OIDC_CALLBACK_ROUTE",
    "OIDC_USERINFO_URL",
)

logger = logging.getLogger(__name__)


class IntrospectTokenValidator(BaseIntrospectTokenValidator):
    """Validates a token using introspection."""

    def introspect_token(self, token_string):
        """Return the token introspection result."""
        oauth = g._oidc_auth
        metadata = oauth.load_server_metadata()
        if "introspection_endpoint" not in metadata:
            raise RuntimeError(
                "Can't validate the token because the server does not support "
                "introspection."
            )
        with oauth._get_oauth_client(**metadata) as session:
            response = session.introspect_token(
                metadata["introspection_endpoint"], token=token_string
            )
        return response.json()


class OpenIDConnect:
    accept_token = ResourceProtector()

    def __init__(
        self,
        app=None,
        credentials_store=None,
        http=None,
        time=None,
        urandom=None,
        prefix=None,
    ):
        for param_name in ("credentials_store", "http", "time", "urandom"):
            if locals()[param_name] is not None:
                warnings.warn(
                    f"The {param_name!r} attibute is no longer used.",
                    DeprecationWarning,
                    stacklevel=2,
                )
        self.accept_token.register_token_validator(IntrospectTokenValidator())
        if app is not None:
            self.init_app(app, prefix=prefix)

    def init_app(self, app, prefix=None):
        # Removed features, die if still there
        for param in _CONFIG_REMOVED:
            if param in app.config:
                raise ValueError(
                    f"The {param!r} configuration value is no longer enforced."
                )
        # Deprecated config values, harmless if still there
        for param in _CONFIG_DEPRECATED:
            if param in app.config:
                warnings.warn(
                    f"The {param!r} configuration value is deprecated and ignored.",
                    DeprecationWarning,
                    stacklevel=2,
                )

        secrets = self.load_secrets(app)
        self.client_secrets = list(secrets.values())[0]

        app.config.setdefault("OIDC_CLIENT_ID", self.client_secrets["client_id"])
        app.config.setdefault(
            "OIDC_CLIENT_SECRET", self.client_secrets["client_secret"]
        )
        app.config.setdefault("OIDC_USER_INFO_ENABLED", True)
        app.config.setdefault("OIDC_INTROSPECTION_AUTH_METHOD", "client_secret_post")
        app.config.setdefault("OIDC_CLOCK_SKEW", 60)
        app.config.setdefault("OIDC_RESOURCE_SERVER_ONLY", False)
        app.config.setdefault("OIDC_CALLBACK_ROUTE", None)

        if "OVERWRITE_REDIRECT_URI" in app.config:
            warnings.warn(
                "The 'OVERWRITE_REDIRECT_URI' configuration value has been replaced by "
                "'OIDC_OVERWRITE_REDIRECT_URI', please update your configuration.",
                DeprecationWarning,
                stacklevel=2,
            )
            app.config.setdefault(
                "OIDC_OVERWRITE_REDIRECT_URI", app.config["OVERWRITE_REDIRECT_URI"]
            )
        app.config.setdefault("OIDC_OVERWRITE_REDIRECT_URI", None)

        app.config.setdefault("OIDC_SCOPES", "openid email")
        if "openid" not in app.config["OIDC_SCOPES"]:
            raise ValueError('The value "openid" must be in the OIDC_SCOPES')
        if isinstance(app.config["OIDC_SCOPES"], (list, tuple)):
            warnings.warn(
                "The OIDC_SCOPES configuration value should now be a string",
                DeprecationWarning,
                stacklevel=2,
            )
            app.config["OIDC_SCOPES"] = " ".join(app.config["OIDC_SCOPES"])

        provider_url = self.client_secrets["issuer"].rstrip("/")
        app.config.setdefault(
            "OIDC_SERVER_METADATA_URL",
            f"{provider_url}/.well-known/openid-configuration",
        )

        self.oauth = OAuth(app)
        self.oauth.register(
            name="oidc",
            server_metadata_url=app.config["OIDC_SERVER_METADATA_URL"],
            client_kwargs={
                "scope": app.config["OIDC_SCOPES"],
                "token_endpoint_auth_method": app.config[
                    "OIDC_INTROSPECTION_AUTH_METHOD"
                ],
            },
            update_token=self._update_token,
        )

        if not app.config["OIDC_RESOURCE_SERVER_ONLY"]:
            app.register_blueprint(auth_routes, url_prefix=prefix)
            app.route("/oidc_callback")(legacy_oidc_callback)
            if app.config["OIDC_CALLBACK_ROUTE"]:
                app.route(app.config["OIDC_CALLBACK_ROUTE"])(legacy_oidc_callback)

        app.before_request(self._before_request)

    def load_secrets(self, app):
        # Load client_secrets.json to pre-initialize some configuration
        content_or_filepath = app.config["OIDC_CLIENT_SECRETS"]
        if isinstance(content_or_filepath, dict):
            return content_or_filepath
        else:
            with open(content_or_filepath) as f:
                return json.load(f)

    def _before_request(self):
        g._oidc_auth = self.oauth.oidc
        if not current_app.config["OIDC_RESOURCE_SERVER_ONLY"]:
            return self.check_token_expiry()

    def check_token_expiry(self):
        try:
            token = session.get("oidc_auth_token")
            if not token:
                return
            if f"{request.root_path}{request.path}" == url_for("oidc_auth.logout"):
                return  # Avoid redirect loop
            token = OAuth2Token.from_dict(token)
            try:
                self.ensure_active_token(token)
            except AuthlibBaseError as e:
                logger.info(f"Could not refresh token {token!r}: {e}")
                return redirect("{}?reason=expired".format(url_for("oidc_auth.logout")))
        except Exception as e:
            logger.exception("Could not check token expiration")
            abort(500, f"{e.__class__.__name__}: {e}")

    def ensure_active_token(self, token: OAuth2Token):
        metadata = self.oauth.oidc.load_server_metadata()
        with self.oauth.oidc._get_oauth_client(**metadata) as session:
            result = session.ensure_active_token(token)
            if result is None:
                # See the ensure_active_token method in
                # authlib.integrations.requests_client.oauth2_session:OAuth2Auth
                raise InvalidTokenError()
            return result

    def _update_token(name, token, refresh_token=None, access_token=None):
        session["oidc_auth_token"] = g.oidc_id_token = token

    @property
    def user_loggedin(self):
        """
        Represents whether the user is currently logged in.

        Returns:
            bool: Whether the user is logged in with Flask-OIDC.

        .. versionadded:: 1.0
        """
        return session.get("oidc_auth_token") is not None

    def user_getinfo(self, fields, access_token=None):
        if not current_app.config["OIDC_USER_INFO_ENABLED"]:
            raise RuntimeError(
                "User info is disabled in configuration (OIDC_USER_INFO_ENABLED)"
            )
        if access_token is not None:
            warnings.warn(
                "Calling user_getinfo with a token is deprecated, please use "
                "g._oidc_auth.userinfo(token=token)",
                DeprecationWarning,
                stacklevel=2,
            )
            return self.oauth.oidc.userinfo(token=access_token)
        warnings.warn(
            "The user_getinfo method is deprecated, please use "
            "session['oidc_auth_profile']",
            DeprecationWarning,
            stacklevel=2,
        )
        if not self.user_loggedin:
            abort(401, "User was not authenticated")
        return session.get("oidc_auth_profile", {})

    def user_getfield(self, field, access_token=None):
        """
        Request a single field of information about the user.

        :param field: The name of the field requested.
        :type field: str
        :returns: The value of the field. Depending on the type, this may be
            a string, list, dict, or something else.
        :rtype: object

        .. versionadded:: 1.0
        """
        warnings.warn(
            "The user_getfield method is deprecated, all the user info is in "
            "session['oidc_auth_profile']",
            DeprecationWarning,
            stacklevel=2,
        )
        return self.user_getinfo([field]).get(field)

    def get_access_token(self):
        """Method to return the current requests' access_token.

        :returns: Access token or None
        :rtype: str

        .. versionadded:: 1.2
        """
        return session.get("oidc_auth_token", {}).get("access_token")

    def get_refresh_token(self):
        """Method to return the current requests' refresh_token.

        :returns: Access token or None
        :rtype: str

        .. versionadded:: 1.2
        """
        return session.get("oidc_auth_token", {}).get("refresh_token")

    def require_login(self, view_func):
        """
        Use this to decorate view functions that require a user to be logged
        in. If the user is not already logged in, they will be sent to the
        Provider to log in, after which they will be returned.

        .. versionadded:: 1.0
           This was :func:`check` before.
        """

        @wraps(view_func)
        def decorated(*args, **kwargs):
            if not self.user_loggedin:
                redirect_uri = "{login}?next={here}".format(
                    login=url_for("oidc_auth.login"),
                    here=quote_plus(request.url),
                )
                return redirect(redirect_uri)
            return view_func(*args, **kwargs)

        return decorated

    def logout(self, return_to=None):
        """
        Request the browser to please forget the cookie we set, to clear the
        current session.

        Note that as described in [1], this will not log out in the case of a
        browser that doesn't clear cookies when requested to, and the user
        could be automatically logged in when they hit any authenticated
        endpoint.

        [1]: https://github.com/puiterwijk/flask-oidc/issues/5#issuecomment-86187023

        .. versionadded:: 1.0
        """
        return_to = return_to or request.root_url
        warnings.warn(
            "The logout method is deprecated, just redirect to {}".format(
                url_for("oidc_auth.logout", next=return_to)
            ),
            DeprecationWarning,
            stacklevel=2,
        )
        return redirect(url_for("oidc_auth.logout", next=return_to))

    def custom_callback(self, *args, **kwargs):
        raise ValueError("This feature has been dropped")
